### DISCUSSION TITLE
Error occurred when getting spikes

### DISCUSSION THREAD

**Discussion by zjm199502** [petersenpeter/CellExplorer#73] [Q&A] [answered]
Hi petersen,

I got a problem when i tried to follow the tutorial. After run the command _`cell_metrics = ProcessCellMetrics('session', session);`_  error occurred as the following picture shows. 

![46757cf5fe84b796c87a14e4888f33d](https://user-images.githubusercontent.com/60344825/194072041-b027bf7d-447c-43c0-aee7-df0896881ee8.jpg)
![9be2100db0750178c14afd75014e3b9](https://user-images.githubusercontent.com/60344825/194072052-f8588194-48b3-441c-a195-17a1fb082548.jpg)

Is there anything wrong?

My raw data are continuous binary data (.dat). And the spike sorter is Spyking-circus.

Thank you very much!

Jinming Zhang

**Comment by zjm199502** [comment]
@petersenpeter

**Comment by petersenpeter** [comment]
Hi

This typically happens if the spike data is not imported correctly (e.g. the spikes struct is empty). Could you check that the spikes struct is generated correctly? I can see that the file `continuous.spikes.cellinfo.mat` was generated but check that the content looks OK? 

If the spikes struct is empty, try generating it directly with `loadSpikes`. 

Feel free to share the session with me.

**Reply by zjm199502** [reply]
Hi,

Does this (Blue Box) looks OK? (I didn't find a right template of `continuous.spikes.cellinfo.mat`)

I tried use `loadSpikes`, and it firstly showed Red Box 1. Then I tried to convert the Spyking-circus sorting result into a Phy format and then `loadSpikes` again, it showed Red Box 2 at this time.   Do I need editor a Spyking-circus result format (.hdf5) by myself ?

![f676555c44679857cc8615fd613f8bf](https://user-images.githubusercontent.com/60344825/194197554-3f029000-4f35-4c0d-974f-fba3ad4a7462.jpg)

Thank you very much!

**Reply by zjm199502** [reply]
And in the fields of `continuous.spikes.cellinfo.mat`, `spikes.times` is empty. While `spikes.processinginfo` and `spikes.processinginfo.params` are as following pictures show.
![Uploading c89e2b3e9be2a0b7aa8bbadf1d17566.jpg…]()
![Uploading 1205dd32b7fa9e339c78c900814882b.jpg…]()

**Comment by petersenpeter** [comment]
Hi. Could you share your dataset with me? I must say I forgot, but Spyking circus has not been implemented yet, but I can fix this today if I have your data.

If you want to import the phy format you must provide these parameters to loadSpikes: 
From the basepath of the session try:
spikes = loadSpikes('format','spyking circus');

Or for the phy data format:
spikes = loadSpikes('format','phy');

**Reply by zjm199502** [reply]
Hi,
The raw data is too big to upload here, so I just send it through email to your address (petersen.peter@gmail.com). Please check the email.

Thank you again for your help.

**Comment by petersenpeter** [comment]
Hi
Sorry, it took me some time to look into this. But I found the issues. 
- Your recording had 32 channels, you had defined 16 in the session info:
session.extracellular.nChannels = 32;

- Same issue with your electrode groups. Create two groups, one for the first bad 16 channels, and a second for the rest:
session.extracellular.electrodeGroups.channels = {[1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16]  [17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32]}; 
This is not strictly necessary but a good practice.

- Next, you did not select the correct spike data format: 
session.spikeSorting{1}.format = 'Phy'

- Finally, your phy data was in a sub-subfolder, continuous/continuous-merged.GUI. Either define the full relative path or move it up a level: 
session.spikeSorting{1}.relativePath = 'continuous/continuous-merged.GUI';

After fixing these issues it worked.

**Reply by zjm199502** [reply]
Hi,
Sorry for the late feedback. 
Thank you so much for your help. And I followed your construction, but unfortunately the error still occurred in my computer when i try to import the phy format result(as mentioned in the first).  
I will try the other spike sorting method that the CellExplorer already supported directly (such as klusta,UltraMegaSort2000). And may I ask you that whether you still have the plan to make Spyking-circus format being supported?

**Reply by petersenpeter** [reply]
Yes, I implement Spyking Circus soon, but as you are doing manual curation in Phy, its actually Phy you want to import, as this contains your manual curation. 

Odd that you still receive an error. What is the error you get? 

Here are the spikes, cell metrics, and session struct for your session that I processed:
[dataset files.zip](https://github.com/petersenpeter/CellExplorer/files/9797458/dataset.files.zip)

**Comment by petersenpeter** [comment, marked as answer]
Hi 
I added SpyKing Circus as a supported format. Please try it for me and let me know if you experience any issues. 

To test it, go to the basepath of your session, and run:
```m
spikes = loadSpikes('format','spyking circus','clusteringpath','continuous','forceReload',true);
```

You can also import your Phy-data from the same dataset:
```m
spikes = loadSpikes('format','phy','clusteringpath','continuous/continuous-merged.GUI','forceReload',true);
```

**Reply by zjm199502** [reply]
Hi,

Thank you so much! First, I can manually curate the sorting results using a matlab-GUI on *.hdf5 files, so it do helps if I can import those directly into CellExplorer. Second, I try the code `spikes = loadSpikes('format','spyking circus','clusteringpath','continuous','forceReload',true);` and it works, spikes were successfully imported.

But there still some issues that I should remind you:
1. In `loadSpikes.m`, line870, file name 'continuous.result.hdf5' actually will change if users modified the raw data file name(*.result.hdf5). 
2. When I try to follow 'CellExplorer_Tutorial.m',  part3.1.1, code `cell_metrics = ProcessCellMetrics('session', session);`, error occurred as following picture shows.
![Picture 1](https://user-images.githubusercontent.com/60344825/196155227-7eaf981a-5141-401b-97e3-1bdf883e7bef.png)

3. After the spike data was imported using the newest code you provide, I try to lanuch the main GUI of CellExplorer using `cell_metrics = CellExplorer('metrics',cell_metrics); `. Error occurred as fllowing picture shows. 
![2](https://user-images.githubusercontent.com/60344825/196155258-4f88b4bc-10cb-456e-85ea-743f43fe8b15.png)

**Reply by zjm199502** [reply]
After `spikes = loadSpikes('format','spyking circus','clusteringpath','continuous','forceReload',true);`, I got three new *.mat files as following shows, without 'continuous.cell_metrics.cellinfo.mat'. Does it need an extra code to generate? 
![3](https://user-images.githubusercontent.com/60344825/196156265-41c78f91-8bae-4b06-8696-5453b36354d2.jpg)

**Comment by petersenpeter** [comment]
1: Thanks for pointing this out: I pushed a new version corrected for this. 
2: I believe I imported the spikes with incorrect dimensions. Try updating your CellExplorer repo and run the script again. 
3: Your third error happens because the cell metrics pipeline resulted in the error in step 2, so you did not create the cell metrics.

**Reply by zjm199502** [reply]
It all work well now. But the code  `spikes = loadSpikes('format','spyking circus','clusteringpath','continuous','forceReload',true);` should be run before `cell_metrics = ProcessCellMetrics('session', session);`

Thank you so much for your time !

Best,
Jinming

ps. software could crash sometime, but I don't know if it should be attributed to my mac, or Matlab. If you haven't meet this, please ignore it.

**Reply by petersenpeter** [reply]
Great!

Regarding crashes. if you think it could be a general thing, send me the errors shown in Matlab! This kind of feedback is vital as I am only able to test it on Mac and Windows, and only with a subset of versions and datasets.

**Reply by zjm199502** [reply]
Crash report has been uploaded, I will try to verfiy this in an another Windows-PC later to see if it is a MacOS issue.  

[Problem report for Matlab.txt.pdf](https://github.com/petersenpeter/CellExplorer/files/9800171/Problem.report.for.Matlab.txt.pdf)

**Reply by zjm199502** [reply]
Crashes seem to be a specific problem related to my mac or matlab version (R2021b). I just try the CellExplorer on my windows-PC (R2022a), it works well and didn't crash anymore. 

Thanks for all of your kind help!

**Reply by petersenpeter** [reply]
Thanks for providing the feedback!

**Reply by zjm199502** [reply]
Hi Dr.petersen,

I'm sorry that I make a mistake that the imported file should be `*.result-merged.hdf5`, which is the right sorting result file that has been mannually curated. I tried to directly change the line870 of `loadSpikes.m`. But then an error occurred again after run `cell_metrics = ProcessCellMetrics('session', session);` as showing below. 

![12](https://user-images.githubusercontent.com/60344825/196381680-79e4ed13-3e8f-44e6-8741-ec2dd1109e15.jpg)
Could please spare some time fixing this?

**Reply by petersenpeter** [reply]
Hi, I fixed some bugs and updated the data source for SpyKing Circus. Update your repo and try to run the pipeline again.

Make sure to delete spikes and cell metrics files before rerunning the pipeline.

**Reply by zjm199502** [reply]
Thank you for that , but the error still occurred. May there be anything still need be modified? 
![321](https://user-images.githubusercontent.com/60344825/196400554-49e85c5d-ee2d-4e12-a536-89c7ce8e123b.png)

**Reply by zjm199502** [reply]
I think I have located the problem: mannually curated `*.result-merged.hdf5` file becomes an 'array' type which has 2 dimensions.
![array](https://user-images.githubusercontent.com/60344825/196436922-c65f12f2-41bd-456f-8520-0577f48e8f5c.png)
 While the uncurated `*.result-merged.hdf5` is 'carray' type which only has 1 dimension. 
![carray](https://user-images.githubusercontent.com/60344825/196436950-df0436ba-0469-4f6f-9232-1f6996d76e13.png)

This could be the reason of the cat error. I try to process the uncurated `*.result-merged.hdf5` file with the latest version of CellExplorer and it does works.

Could this problem be solved by modifying current script? Or I need turn to the developer of Spyking-circus for further helping?

Sorry for so many questions.

**Reply by zjm199502** [reply]
I made some modifications on `loadSpikes.m` so that it can correctly handle the mannually curated `*.result-merged.hdf5` with 'array' data type. Hope this will help the others in using a Spyking-circus format.
Replace code line882 `spikes.times{i} = double(h5read(nwb_file,['/spiketimes/',template_names{i}]))/session.extracellular.sr;`
with:
`times = h5read(nwb_file,['/spiketimes/',template_names{i}]);`
`spikes.times{i} = double(times(1,:)')/session.extracellular.sr;`

Thank you so much!

**Reply by petersenpeter** [reply]
Great. I will incorporate your suggestion and commit a new version of loadSpikes.
